### DISCUSSION TITLE
include a matplotlib animation

### DISCUSSION THREAD

**Discussion by stoffprof** [executablebooks/meta#661] [Q&A] [answered]
Has anyone had success including an HTML animation from `matplotlib` in a book?

For example, I'm trying to do something like what's on [this web page](https://matplotlib.org/stable/gallery/animation/simple_anim.html): embed an animation directly in the HTML. The approach on that page looks like it involves Sphinx, but I'm not sure how they're doing it.

Would appreciate any ideas if you've done something like this.

**Comment by choldgraf** [comment, marked as answer]
Hmmm - it looks like they're using `sphinx-gallery`, and specifically the animation is handled by this code:

https://github.com/sphinx-gallery/sphinx-gallery/blob/9245f7ce482c5e26d59e27caf9f5bfcbc863ffb2/sphinx_gallery/scrapers.py#L217-L241

Maybe that could be adapted somehow for your use-case?

**Reply by stoffprof** [reply]
Thanks! I'll keep poking around in that code but I'm having a hard time seeing what the workflow would look like to do this. Any additional tips would be much appreciated.

**Reply by choldgraf** [reply]
Does the animation work in a jupyter notebook interface? If so, what happens when you build your book with it?

**Reply by stoffprof** [reply]
I'm generating the animation as HTML so it does not work directly in jupyter lab. The output is a static picture both in jupyter lab and in the generate book.

I create an animation by calling my `animate` function:

`anim = animation.FuncAnimation(fig, animate, frames=201, blit=True)`

I'm then saving it as HTML like this:

```
writer = animation.HTMLWriter(fps=5)
anim.save("animation.html", writer=writer)
```

This just creates an HTML file and a folder with a bunch of images that the HTML uses javascript to refer to to generate the animation. Opening the file in my browser works.

I thought I could somehow embed this HTML in my book, or get matplotlib to generate something that Sphinx could use, but maybe this isn't a good approach. I'm happy to try something different if you have any ideas. I'm somewhat novice at this, so appreciate the handholding. I'm just not seeing how to piece these things together.

Thanks!

**Reply by choldgraf** [reply]
hmmm could you try to just embed the HTML directly? You could try a block like:

````
```{raw} html
<div>your html</div>
```
````

**Reply by stoffprof** [reply]
I think it's just ignoring it. The HTML has a bunch of javascript in it, so I can't stick it in a `<div>` tag. It's like this:

```
```{raw} html
<link rel="stylesheet"
href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
<script language="javascript">
  function isInternetExplorer() {
...etc...
```

I am able to embed the file using a raw `<iframe>` tag in the notebook, and that works, but it doesn't seem like the best solution.

**Reply by choldgraf** [reply]
How about:

````
```{raw} html
:file: yourfile.html
```
````
?

**Reply by stoffprof** [reply]
That worked! Thanks so much. So simple.

**Reply by choldgraf** [reply]
any interest in adding to the documentation here: https://jupyterbook.org/advanced/html.html#use-raw-html-in-markdown

?

I think that section is too-discouraging of HTML. It doesn't document the use of `{raw}`, nor the way to include a file via `:file:`. It'd be helpful to others if that information were there!

**Reply by stoffprof** [reply]
Sure, although I have a pretty dumb question first: where do the source files for the documentation live?

**Reply by choldgraf** [reply]
Here: https://github.com/executablebooks/jupyter-book/edit/master/docs/advanced/html.md

Although you can also find a link in the little github icon at the top right:

<img width="407" alt="Slack_6ztK4PVVW9" src="https://user-images.githubusercontent.com/1839645/153281033-c10bbcde-cc1b-4321-9b2f-f363be92d231.png">
